### PR TITLE
Fix S4158 FP: The analyzer incorrectly concludes the HashSet is guaranteed to be empty

### LINKED ISSUE #7582: Fix S4158 FP: rule raised where there is no guarantees of emptiness
### Description

S4158 regarding iteration of known empty collections may be raised in situations where there is no guarantee that the collection is empty.

### Repro steps

```cs
private static void Demo(IEnumerable<int> input)
{
  var list = new List<int>();
  var flag = false;

  foreach(var item in input)
  {
    if (item > 10)
    {
      flag = true;
      continue;
    }

    list.Add(item);
  }

  if (flag)
  {
    foreach(var item in list)
    {
      // do something with items that were <= 10
      // only when items > 10 were found.
    }
  }
}
```

### Expected behavior

S4158 is not raised, because there is no guarantee that `list` will be empty when `flag` is true.

### Actual behavior

S4158 is raised.

### Known workarounds

Suppress the rule.

### Related information

* C#/VB.NET Plugins version : C# 9.4.0.72892 (SonarLint  7.0.0.74072)
* Visual Studio version: 17.6.4
* MSBuild / dotnet version: dotnet 7.0.304
* Operating System: Windows 11

### LINKED ISSUE #8041: Fix S4158 FP: The analyzer incorrectly concludes the HashSet is guaranteed to be empty
### Description

HashSet.Remove causes [S4158](https://sonarsource.github.io/rspec/#/rspec/S4158/csharp) on an empty HashSet. It should not.

### Repro steps

I tried to remove parts of this repro, but I could not make it cleaner without loosing the FP. `_ = toDoLater.Remove(item);` issues  [S4158](https://sonarsource.github.io/rspec/#/rspec/S4158/csharp)

```
    public interface IFoo<T>
    {
        public bool someProperty { get; }

        public IEnumerable<T> Successors { get; }
    }

    public static void Func<T>(this ICollection<T> toDo)
        where T : class, IFoo<T>
    {
        Queue<T> toDoNow = new();
        HashSet<T> toDoLater = new();

        foreach (T item in toDo)
        {
            if (item.someProperty)
            {
                toDoNow.Enqueue(item);
            }
            else
            {
                _ = toDoLater.Add(item);
            }
        }

        while (toDoNow.Count > 0)
        {
            T current = toDoNow.Dequeue();

            // Handle current

            foreach (T item in current.Successors)
            {
                _ = toDoLater.Remove(item);
                toDoNow.Enqueue(item);
            }
        }
    }
```

### Expected behavior

[S4158](https://sonarsource.github.io/rspec/#/rspec/S4158/csharp) is not issued as the HashSet is not guaranteed to be empty

### Actual behavior

[S4158](https://sonarsource.github.io/rspec/#/rspec/S4158/csharp) is issued.

### Known workarounds

Suppress the issued warning.

### Related information

SonarAnalyzer version 9.7.0.75501

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Fixes #8041
Fixes #7582

We recently improved how the SE engine handles loops (#7795). We missed to include foreach loops in the process.

### CODE REVIEW

**Review by mary-georgiou-sonarsource — CHANGES_REQUESTED** [review]
Please add a reprocude also for the issue itself
https://github.com/SonarSource/sonar-dotnet/issues/8041

**Comment by Tim-Pohlmann** [comment]
@mary-georgiou-sonarsource is this necessary? There are multiple similar UTs in the code base already, which are getting fixed by this PR.

File: analyzers/src/SonarAnalyzer.VisualBasic/SymbolicExecution/Roslyn/VisualBasicSyntaxClassifier.cs [hunk 1]
```diff
@@ -37,6 +37,7 @@ node.Parent switch
             DoStatementSyntax doStatement => doStatement.WhileOrUntilClause,
             LoopStatementSyntax loopStatement => loopStatement.WhileOrUntilClause,
             ForBlockSyntax forBlock => forBlock.ForStatement,
+            ForEachStatementSyntax foreachStatement => foreachStatement.Expression,
```
Comment (MaStr11) [inline analyzers/src/SonarAnalyzer.VisualBasic/SymbolicExecution/Roslyn/VisualBasicSyntaxClassifier.cs:40]: Misses https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.foreachvariablestatementsyntax?view=roslyn-dotnet-4.6.0

e.g `foreach(var (i, j) in new (int, int)[0])`

Should be `CommonForEachStatementSyntax` instead.

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**